### DISCUSSION TITLE
Workflow Parameter Definitions

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#246] [Design & Development]
This discussion topic is a followup from ideas that emerged from a related discussion page on parameter verification.    The issue here is if we could design build a data structure that could encapsulate all the input parameteric data for a workflow?   

I would assert that the answer is yes if we limit the universe to things that can be saved in a mongodb document.  Given our experience to date, that is pretty all encompassing.  The only thing I can see it would not cover are large auxiliary data objects that wouldn't fit in a document itself and would require storage in something like gridfs or a set of files.   A type example from the workflow that has driven this question is 3D earth models that are used by the plane wave migration program I have been adapting to mspass.  Any 3D model is large enough it would be unlikely to fit in a single mongodb document, which online sources say is 16 Mb.   

That does, however, mean all of the following could be captured in a  mongodb document:

- any simple parameters passed as function arguments expressed as key-value pairs defining simple types
- any parametric data that could be expressed as a python dict (or mspass Metadata)
- the AntelopePf class could be easily stored as a document (I think there may even be code to do this from earlier in the development)
- pickled python objects; we did this with Inventory and Catalog objects already but any data that could be expressed as a python class could be stored that way
- There may be others; I haven't reviewed the mongodb bson types

So you would this be done?   I suggest the entire parameter set for a workflow could be expressed in a document with two key features:
1.  The top level structure of the document is keyed by algorithm name + instance.   That would map perfectly to our global history concept.   (e.g. one might use filter_1 for instance 1 of the filter algorithm.)   An alternative would be subdocument under algorithm keyed by an "instance" tag of some sort.  (e.g. { 'filter' : { '1' : {'arg0' : 'highpass','corners' : 2.0}},{'2' : {'arg0' : 'lowpass' :'corners' : 3.0}} - noting "arg0" is used here as a way to specify an position dependent parameter that doesn't require a key.   
2. Parameters would be subdocuments of whatever level required to express the hiearchy of the parameters used as input to the function and instance with which they are associated.  In my example a simple example is {'arg0' : 'highpass', 'corners' : 2.0}, but it could be a lot more elaborate with levels of subdocuments.   

Let me post this to get us started.   Externally I'll build an example of what might be used as input for a pwmig workflow.   When I finish that I'll post it below this one.

**Comment by pavlis** [comment]
Here is an example cobbled from  subset of what is needed for pwstack and pwmig.  

Suppose we had this as a parameter file for an algorithm called pwstack and we described it in pf format as follow:
```
stack_count_cutoff 5
centroid_cutoff 20.0
data_window_start_time -5.0
data_window_end_time 180.0
data_top_mute &Arr{
  end_time 1.0
  zero_end_time 2.0
}
depth_dependent_aperture &Tbl{
  0.0 75.0 150.0
  700.0 75.0 150.0
}
```
pwstack has a sibling algorithm called "pwmig".  Suppose we had this for its parameter file:
```
dt 0.05
10
use_grt_weights true
Pslowness_model mit2011
```
Noting as a warning to any readers of this in the distant future when pwstack and pwmig should be in working order that these are NOT actual complete parameter sets for either algorithm.  They are a subset that illustrate all the types of constructs in an antelope pf;  ie. simple name-value pairs, Arr, and Tbl.  

the above could be put in a single json file (Note three is definitely a syntax error in the curly brackets below that I can't quite figure out, but I trust it show the idea) something like the following that could be directly loaded into MongoDB:
```
{
'pwstack' : {
  {  '1' :  {
      'stack_count_cutoff' : 5,
      'centroid_cutoff' : 20.0,
      'data_window_start_time' : -5.0,
      'data_window_end_time' : 180.0,
      'data_top_mute' : {'end_time' : 1.0,'zero_end_time' : 2.0 },
      'depth_dependent_aperture' : ['0.0 75.0 150.0','700.0 75.0 150.0']
     }
  }
},
'pwmig' : {
  { '1' :  {
      'dt' : 0.05,
      'border_padding' : 10,
      'use_grt_weights' : True,
      'Pslowness_model' : 'mit2011'
    }
  }
}
}
```
Noting this example uses the algorithm names as the top level keys (here *pwmig* and *pwstack*) and instance is used as a secondary key to map into our history concepts.  The idea is that if the above were a mongodb document assigned the symbol *doc* you could retrieve the parameters from instance "1" of pwmig as:
```
params=doc['pwmig']['1']
```
As I said I can't seem to get the syntax right, but I think that is possible.  I created similar things with python code, but can't get seem to get the json syntax quite right.  I hope this shows the idea

**Comment by pavlis** [comment]
I'm getting back to this again and am working on implementing the following:
```
def pf_dbsave(db,pf,name_tag, collection='AntelopePf'):
    """
    Save an AntelopePf object to MongoDB as document. If

    The data structure of an AntelopePf is actually not dependent upon the
    format invented by Dan Quinlan of BRTT they call a "pf file".
    A Pf basically defines a tree structured set of data indexed with key-value
    pairs.  That maps exactly into a MongoDB document when subdocuments of
    any level are allowed.  The one item in a Pf that does not mesh perfectly
    is a Tbl which in the API is converted to an std::list of strings.  The
    consumer of the data in a Tbl is responsible for parsing the data in that
    list of strings.

    This function saves an AntelopePf as a document in a collection defined by
    the collection argument.   No schema is enforced on what is saved
    the AntelopePf is strongly typed.

    :param db:  MsPASS database handle (a parent MongoDB database handle should work also)
    :param pf:  The AntelopePf object that is to be saved.
    :param name_tag:   This is required to be a string that is saved with
      the frozen key name of 'pfname'.   (e.g. if name_tag is 'pwmig' if
      if successful this function will create an entry {'pfname' : 'pwmig'}
      in the document stored by that call to this function).  The AntelopePf
      must not contain the keyword pfname.
    :param collection:  MongoDB collection to which the data should be stored.
      (default is 'AntelopePf')

    :return: ObjectID of the document saved.
    """
    # First we do a few sanity checks.
    if not isinstance(name_tag,str):
        raise MsPASSError('pf_dbsave:  name_tag argument must be a string','Fatal')
    # All the work is done in this function.  It returns a dict but we call it
    # doc only because we will treat it that way when we save it to Mongodb
    doc = pf_to_dict(pf)
    # We can't allow doc to contain the keyword pfname
    if 'pfname' in doc.keys():
        raise MsPASSError('pfsave:  pf object contains a value for the keyword=pfname.  Not allowed','Fatal')
    doc['pfname'] = name_tag
    col = db[collection]
    result = col.insert_one(doc)
    return result.inserted_id

def pf_dbread(db,doc_or_id):
   """
   Read data from MongoDB and return the document defined by doc_or_id as an AntelopePf object.   
  """
def pf_to_dict(pf):
    """
    Returns a dict representation of an AntelopePf object.  Simple parameters
    map directly to key-value pairs.  Branches (Arr& in the pf format) are
    converted to dict values keyed to the Arr key name.   (Note the conversion
    is recursive so branches of artibrarily deep level can be converted.)
    This function can be called standalone for a conversion but is used
    by the pf_dbsave function as a first step because a dict can be saved
    directly in MongoDB with pymongo.

    :param pf:  AntelopePf object to be converted.
    :return: dict that is an alternate representation of pf
    """
def dict_to_pf(dict):
    """
    Inverse of pf_to_dict.
    """
```

The hard part of this the two converters:  `pf_to_dict` and `dict_to_pf`.    The key point for discussion is if there are other ways to create parameter trees that we could pretty much support by the same mechanism?

**Reply by pavlis** [reply]
The process of actually trying to implement the ideas suggested above led me realize I had made a fundamental error:  I forgot the fundamental, modern principle of abstraction.   In particular, I was forgetting that the external representation of something should not define the thing we are trying to define as a generic object.   This is a case in point because the Antelope "pf" format and equivalently a yaml file are external representations of a common concept.  That common concept is a tree structure storage of set of key-value pairs used to define parametric (control) parameters for a workflow.  

With that I propose we define a WorkflowParameters class that we use to hold parametric data for any processing algorithm in MsPASS.   (name is a strawman proposal - totally open to alternative suggestions)   Some design features of WorkflowParameters:
1.  The basic constructor should be a MongoDB document.   I am pretty certain MongoDB documents with arbitrarily deep subdocuments (dict of dicts in pymongo) can represent in parameter tree we can conceive of at this point that could support our users.
2.  The api should be fundamentally different from AntelopePf as it needs to abstract the concept to the generic one of a tree structure of name-value pairs instead of the more procedural approach from which AntelopePf was born.   (None of you are likely to realize this was born from a procedure plain C library written in the early 1990s).  There are almost certainly many models of this in the xml world.   That is, I've seen parsers for xml that abstract an xml file as a tree structure.   In any case, the core api needs to be thought through.
3. We should have functions that convert our generic WorkflowParameter class to supported parameter types.  Ones I recall are:  (a) function arguments, (b) AntelopePf, (c) Metadata, and (d) (not previously supported but perhaps time) yaml export to the structure used in the yaml reader we are using.   
4. We may need adapters to mesh this with our global history manager - the reason we are going through this exercise.  

We should add discussion of this topic to the agenda of our group meeting this week.

**Reply by wangyinz** [reply]
Yeah, this abstraction is a great idea. I think this `WorkflowParameter` can be simply derived from a plain dictionary - it is the basic class that supports the tree structure of key-value pairs you describe above. 

Currently we are implementing support for pf and yaml files in the global history manager, but I don't think we have the design to integrate everything into a single object as you suggested (although they will be recorded in the database as a single document anyway). I think it make sense to create such generic class for the purpose. 

One thing that I think need clarification here is the item (c) in point 3. I am not sure what Metadata is referring to here, but I guess it is not the Metadata of the data object (i.e., TimeSeries or Seismogram), but I am not sure how Metadata can be used in a workflow though. We don't have a function or algorithm that uses it as parameter. Do you have some examples in mind?

**Reply by pavlis** [reply]
Thanks for the comments.   I think I will now aim to flesh this out a bit more. 

As to the question about Metadata.  What I was thinking was something simple.  If an application needed only key-value pairs it could be encapsulated in a Metadata container without the additional baggage of this new WorkflowParameter.   It probably would be better restated as py::dict/Metadata since they are are variants of a common concept.

**Reply by wangyinz** [reply]
Gotcha! I guess we need to have examples of application/functions that takes input parameters as 1) arguments, 2) files (pf, yaml, or something equivalent) and 3) Metadata. Right now, I think we only have 1) and 2), and for 2) it is only the decon function that takes pf files as parameters (or it can be treated as a mix of 1 and 2). We probably need more examples and these may involve mixing the three types.

**Comment by pavlis** [comment]
Made a first attempt at this.   First, consider the functionality we need in a base class = core concepts:
1.  Store key-value pairs in a G-tree structure.   That is, I think, the generic name for this.  The gneric concept is described, for example, [here](https://www.cs.cmu.edu/~clo/www/CMU/DataStructures/Lessons/lesson4_1.htm).  Leaves are simple name-value pairs of parameters.  Branches would be respresented externally by subdocuments in MongoDB
2.  We need simple methods to climb up or down the tree.
3.  We need to be able to extract a branch from the tree at any level to provide clean hierarchy of parameters.  For instance, below is a suggestions for a special tree (child of the base class) where the root of the tree is only a set of branches to each algorithm:alg_id in a workflow.   

That said, here is a suggested starting point for a more abstract base class:
```
class ParameterGTree:
  """
  Base class for family of objects used to hold an abstraction of a
  set of control parameters for a data processing function.   The
  base class abstracts the concept of storing such data in an g-tree
  structure.  In the documentation here the tree should be pictures as
  upright in the biological tree analog. i.e. up mean higher levels in
  the tree and down means dropping to lower levels in the tree.

  Each node of the g-tree may have leaves and/or a set
  of branches.   Most simple algorithms need only one node with leaves
  made up of name-value pairs.  More complex algorithms often need
  the tree structure to describe more complicated data control structures.
  """
  def __init__(self,doc):
    """
    Construct from a MongoDB document, which with pymongo is equivalenced
    to a python dict.  Branches are defined by subdocuments.
    """
    self.control = doc

  def leaf_keys(self):
    """
    Return the keys for all key-value pairs that are leaves at the current
    level of the parameter tree.   For branches this method can be used
    to extract leaves at the current level. Return is a dict of key-value
    pairs that we are calling the leaves of the tree.
    """
  def branch_keys(self):
    """
    Return the keys for all branches from this level.   Branches are keyed with a keyword
    string like leaves.  Branches can be extracted with prune or we can
    walk the tree with a set of methods defined below.
    """
  def get_branch(self,key):
    """
    Extract the contents of a named branch.  Returns a copy of the tree
    with the associated key from the branch name upward.  The tree
    returned will have the root of the tree set as current.
    """
  def prune(self,key):
    """
    Remove a branch defined by key from self.  Return a copy of the
    branch pruned in the process (like get_branch but self is altered)
    """

  def get(self,key,seperator='.'):
    """
    Fetch a value defined by key.  For leaves at the root node the
    key can be a simple string.  For a leaf attached at a higher level node
    we specify a chain of one or more branch names with keys between
    the specified seperator.  Examples (all used default value of seperator):
    1.  If we had a leaf node with the key 'name' under the branch name
        'phases' we use the compound key 'phases.name'. Such a tag could, for
        example, contain 'P' for the seismic to define this set of parameters
        as related to the P phase.
    2.  Suppose the phases branch was linked to higher level branch with
        the key 'travel_time_calculator' that had a leaf parameter 'taup'
        that is itself a branch name with terminal leaf keys under it.
        A real life example might be 'model_name'.  We would refer to that
        leaf with the string 'phases.travel_time_calculator.taup.model_name'.
        That key might, for example, have he value 'iasp91' which could be
        passed to obspy's taup calculator.

    This method should also support key defined as a python list.   The list
    would need to be a set of keys that would define the path to climb the
    tree to fetch the desired leaf.  For this form the two examples above
    would be represented as follows:
    1.  ['phases','name']
    2.  ['phases','travel_time_calculator','taup','model_name']
    """
  def put(self,key,separator='.'):
      """
      putter with same behavior for compound keys defined for get method.
      A put should create a new branch it implies if that branch is not
      already present.
      """
  def sprout(self,key,seperator='.'):
    """
    Add an empty branch with tag key. Compound keys are as described in
    get method above.
    """
```
Something I considered for this base class, but decided belonged in a subclass if desired was a tree where the object had an internal pointer to a particular level of the tree as "current".   A familiar and very old example is "current working directory" of the unix shell.   Incidentally, it just dawned on me that the very concept here could be conceptualized as a variant of the unix file system:   the leaves of the trees are file names and the directories are the nodes of the tree.  The only difference here is that leaves can contain more than file names.   In my opinion, however, that concept does no belong in the base class.  

The other thing to stress is the suggested (nongeneric) use here of compound keys to simplify the process of climbing up the tree from the root.   Using the file system analogy that is equivalent to allowing full path names as keys with "/" as the separator for unix file names and "\" for Windows file names.  For consistency with MongoDB "." is the right default separator because that syntax for subdocuments is supported to some extent - e.g. I think you can use that syntax in a find query but I may remember this wrong;  just know they use that for subdocuments in som contexts. 

All team members.  Critically consider: (a) are all of these methods core or should some be viewed as specializations, and (b) are there missing methods that should be thought of as core?

With that design we have most of what we need and the following would be my suggestion for a prototype design for a tree we could use in our global history manager:
```
class WorkflowParameters(ParameterGTree):
  def __init__(self,doc):
    """
    Construct from a MongoDB document that is equivalent to a python dict.
    Concept here is the external representation of this object is that one
    document (with multiple branches = subdocumnts) extracted by a find
    operation with MongoDB.
    """
  def algorithm_keys(self):
     """
     Return the set of keys at the root of the tree that are algorithm
     names used in a workflow.  We may want this method to return
     compound keys or a list of pairs (2 element tuples) of
     algorithm:alg_id values that defin a unique set of parameters.
     """
  def algorithm_parameter_type(self,key):
    """
    This method should provide a means to define the type of parameter set
    a particular algorithm expects (alg_id should not be required here.  A
    unique algorithm (function) should always use the same type of parameters)
    The type should be thought of as a literal object type that python
    could test with isinstance.   Conversion procedures should use this
    method to know how to parse the contents.  I think for now we could
    probably return one of three things:  'arguments', 'ParameterGTree', or 'text'
    'arguments' would be a dict of name-value pairs, which is a subset of
    a ParameterGTree.   I think anything else we know of can be captured by
    these three.  A ParamterGTree is generic and we should have a converter
    for an AntelopePf.   The 'text' type would provide a way to easily retrofit
    old school programs that want to read a file.
    """
  def get_arg_parameter(self,key):
    """
    Return a dict of parameters for a function run with parameters that
    can be expressed as name-value pairs.
    """
  def get_parameter_tree(self,key):
    """
    Return a ParameterGTree encapsulation of the data for algorithm defined by
    key.  This method would likely be implemented as a get_branch using
    key.   This method maybe should have the signature (self,algorithm, alg_id).
    """
  def get_text_file_handle(self,key):
    """
    Return a file like object handle to a text segment that the caller can
    parse.
    """
```
The key idea here as noted in my draft docstrings is that parameter input to an algorithm can be abstracted as one of three forms:  (a) an argument string of simple (i.e. low level language constructs of float, int, bool, and string/char*), (b) hierarchic parameters that could themselves be expressed as the base class ParameterGTree object, and (c) what I've called a text file.   Note an AntelopePf is mostly (b) but includes concepts of (c).   What I mean by a text file is that many legacy codes input all the data they need via a set of auxiliary files.   That can be as simple as a small table of numbers to represent something like a layered earth model or some complicated input data file structure that the program parses with a specialized set of parsers.   The later was the norm in the old Fortran days and was common in early procedural C programs.  Still is the standard approach by most old guys like me.  In any case, we need a clean abstraction of text file data to simplify conversion of legacy code.   There is a caution about text data though.   I am 100% sure large text files stored that way would be trouble because MongoDB has an upper limit on the size of a single document of a few megabytes.   Things like 3D earth models as "parameters", for example, are a bad idea.  We would need do document clearly that large data files should be defined by a reference with a file name parameter and not stored inside documents. 

I suggest the idea of this abstraction is we internally use these data objects but provide convenience procedures to interact with the class.  An example is a converter to convert a ParameterGTree to an AntelopePf.   I think the actual implementation can be a python dict with branch nodes defined by keys where the value is another python dict.  Higher level branching just adds another level of key indexing to more dict containers.   Anyone implementing a new algorithm and familiar with python would likely want to just use that structure directly.  

One final point is working with trees like this will almost inevitably involve using recursive functions for some algorithms.   As you all may know recursion is dangerous and we should not encourage it except where essential.

**Reply by pavlis** [reply]
Postscript to above: I noticed in reviewing this that you will see a few relics of my initial thoughts of supporting the idea of a "current level".   Any reference to "current" in the docstrings should now be translated to "root" as the idea in the abstraction is you would use get_branch if you wanted a subset of the tree to be passed to a particular algorithm.
